doc/bsdsum.1
.\"	$OpenBSD: cksum.1,v 1.39 2016/09/03 17:01:01 tedu Exp $
.\"
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Copyright (c) 2022
.\"	Nicolas Provost
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)cksum.1	8.2 (Berkeley) 4/28/95
.\"	@(#)bsdsum.1	2022/12/11
.\"
.Dd $Mdocdate: December 11 2022 $
.Dt BSDSUM 1
.Os
.Sh NAME
.Nm bsdsum
.Nd display file checksums and block counts 
.Sh SYNOPSIS
.Nm bsdsum
.Bk -words
.Op Fl bcpqrtx
.Op Fl a Ar algorithms
.Op Fl C Ar checklist
.Op Fl h Ar hashfile
.Op Fl s Ar string
.Op Ar
.Ek
.Sh DESCRIPTION
The
.Nm bsdsum
utility writes to the standard output a single line for each input file.
The format of this line varies with the algorithm being used as follows:
.Bl -tag -width allxothers
.It bsdsum
The output line consists of three whitespace separated fields:
a CRC checksum, the number of octets in the input,
and name of the file or string.
If no file name is specified, the standard input is used and no file name
is written.
.It all others
The output line consists of four whitespace separated fields:
the name of the algorithm used, the name of the file or string in
parentheses, an equals sign, and the cryptographic hash of the input.
If no file name is specified, the standard input is used and only
the cryptographic hash is output.
.El
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl a Ar algorithms
Use the specified algorithm(s) instead of the default (md5).
Supported algorithms include md5, sha1, sha224, sha256, sha384, sha512.
Multiple algorithms may be specified, separated by a comma or whitespace.
Additionally, multiple
.Fl a
options may be specified on the command line.
Case is ignored when matching algorithms.
The output format may be specified on a per-algorithm basis
by using a single-character suffix, e.g.\&
.Dq sha256b .
If the algorithm has a
.Sq b
suffix, the checksum will be output in base64 format.
If the algorithm has an
.Sq x
suffix, the checksum will be output in hex format.
If an algorithm with the same output format is repeated,
only the first instance is used.
.It Fl b
Output checksums in base64 notation, not hexadecimal by
default.
A
.Sq b
or
.Sq x
suffix on the algorithm will override this default.
.It Fl C Ar checklist
Compare the checksum of each
.Ar file
against the checksums in the
.Ar checklist .
Any specified
.Ar file
that is not listed in the
.Ar checklist
will generate an error.
.It Fl c
If this option is specified, the
.Ar file
options become checklists.
Each checklist should contain hash results in the normal format,
which will be verified against the specified paths.
Output consists of the digest used, the file name,
and an OK, FAILED, or MISSING for the result of the comparison.
This will validate any of the supported checksums.
If no file is given, stdin is used.
The
.Fl c
option may not be used in conjunction with more than a single
.Fl a
option.
.It Fl h Ar hashfile
Place the checksum into
.Ar hashfile
instead of stdout.
.It Fl p
Echoes stdin to stdout and appends the
checksum to stdout.
.It Fl q
Only print the checksum (quiet mode) or if used in conjunction with the
.Fl c
flag, only print the failed cases.
.It Fl r
Reverse the format of the hash algorithm output, making
it match the checksum output format.
.It Fl g
Output the digest in GNU coreutils format.
.El
.Pp
The default CRC used is based on the polynomial used for CRC error checking
in the networking standard
ISO/IEC 8802-3:1996.
The other available algorithms are described in their respective
man pages in section 3 of the manual.
.Sh EXIT STATUS
.Ex -std bsdsum
.Sh SEE ALSO
.Xr md5sum 
.Xr sha1sum 
.Xr sha256sum 
.Xr sha384sum 
.Xr sha512sum 
.Xr openssl 
.Pp
The default calculation is identical to that given in pseudo-code
in the following ACM article:
.Rs
.%T "Computation of Cyclic Redundancy Checks Via Table Lookup"
.%A Dilip V. Sarwate
.%J "Communications of the ACM"
.%D "August 1988"
.Re
.Sh STANDARDS
The
.Nm
utility is compliant with the
.St -p1003.1-2008
specification.
.Pp
All the flags are extensions to that specification.
.Sh HISTORY
The
.Nm bsdsum (cksum)
utility appeared in
.Bx 4.4 .
.Sh CAVEATS
Do not use the md5 algorithm to verify file integrity.
An attacker can trivially produce modified payload that
has the same checksum as the original version.
Use a cryptographic checksum instead.


ChangeLog
bsdsum
======

2022/12, 2023/01 - v1.4
* add split digest
* add binary style
* add configure script
* rewrite filelist parser
* add gcc or clang support
* add tests
* fix BSD support

2022/12/18 - v1.3
* use mmap to access content of files
* add -f (offset) and -l (length) options

2022/12/17 - v1.2
* add SHA3-256 and SHA3-512 support
* add again auto-test option [-t]
* add -s STYLE option
* change -h to help and old -h to -o
* change various options and flags

2022/12/11 - v1.1
* add libcrypto files from libressl
* add -g option (GNU output format)
* remove some options



src/bsdsum.c
/*
 * Copyright (c) 2022-2023
 *      Nicolas Provost <dev AT npsoft DOT fr>
 * Copyright (c) 2001,2003,2005-2007,2010,2013,2014
 *	Todd C. Miller <millert@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 * Sponsored in part by the Defense Advanced Research Projects
 * Agency (DARPA) and Air Force Research Laboratory, Air Force
 * Materiel Command, USAF, under agreement number F39502-99-1-0512.
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <sys/mman.h>
#include <sys/wait.h>
#include <ctype.h>
#include <err.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <time.h>
#include <unistd.h>
#include <errno.h>
#include <signal.h>

#include "bsdsum.h"

static void bsdsum_init (bsdsum_t *bs)
{
	memset(bs, 0, sizeof(bsdsum_t));
	bs->length = -1;
	bs->offset = -1;
	bs->style = STYLE_DEFAULT;
	bs->data = calloc(BUFFER_SZK, 1024);
	bs->ofile = -1;
	bs->enc64 = ENC64_NONE;
}


/* copy 'hf' as tail of the list bs->hl and return it */
static bsdsum_op_t* bsdsum_add_op (bsdsum_t *bs, bsdsum_op_t *hf, 
					bsdsum_enc64_t enc64)
{
	bsdsum_op_t *hftmp;
	bsdsum_op_t *o;

	hftmp = calloc(1, sizeof(*hftmp));
	if (hftmp == NULL)
		err(1, "out of memory");
	*hftmp = *hf;
	if (enc64 != ENC64_NONE) {
		hftmp->base64 = 1;
		hftmp->enc64 = enc64;
	}
	hftmp->next = NULL;
	if (bs->hl == NULL)
		bs->hl = hftmp;
	else {
		for (o = bs->hl; o && o->next; o = o->next) { }
		o->next = hftmp;
	}
	return hftmp;
}

/* count the number of operations in bs->hl */
static int bsdsum_count_op (bsdsum_t *bs)
{
	bsdsum_op_t *o;
	int n;

	for (n = 0, o = bs->hl; o; o = o->next, n++) { }
	return n;
}

int main (int argc, char **argv)
{
	bsdsum_t bs;
	bsdsum_op_t *hf, *hftmp;
	char *cp;
	const char *optstr = "a:C:co:hpts:f:l:";
	char* endptr;
	int fl, i;
	int use_split = 0;
	int style_spec = 0;
	int a_opts = 0;

	bsdsum_init(&bs);

	/* Args pass 1 */
	while ((fl = getopt(argc, argv, optstr)) != -1) {
		switch (fl) {
		case 'h':
			bsdsum_help();
			exit(0);
			break;
		case 's':
			if (strcasecmp("base64", optarg) == 0) {
				bs.style = STYLE_BASE64;
				bs.enc64 = ENC64_DEFAULT;
			}
			else if (strcasecmp("sym64", optarg) == 0) {
				bs.style = STYLE_BASE64;
				bs.enc64 = ENC64_SYM;
			}
			else if (strcasecmp("gnu", optarg) == 0)
				bs.style = STYLE_GNU;
			else if (strcasecmp("cksum", optarg) == 0)
				bs.style = STYLE_CKSUM;
			else if (strcasecmp("terse", optarg) == 0)
				bs.style = STYLE_TERSE;
			else if (strcasecmp("binary", optarg) == 0)
				bs.style = STYLE_BINARY;
			else if (strcasecmp("default", optarg)) {
				bsdsum_help();
				exit(1);
			}
			style_spec = 1;
			break;
		case 'l':
			bs.length = strtol(optarg, &endptr, 0);
			if ((bs.length < 0) || (endptr && *endptr))
				errx(1, "bad value for -l");
			break;
		case 'f':
			bs.offset = strtol(optarg, &endptr, 0);
			if ((bs.offset < 0) || (endptr && *endptr))
				errx(1, "bad value for -f");
			break;
		case '?':
			bsdsum_help();
			exit(1);
		}
	}
	if (bs.length >= 0 || bs.offset >= 0) {
		if (style_spec && bs.style != STYLE_TERSE)
			errx(1, "-f/-l implies -s terse");
		bs.style = STYLE_TERSE;
		bs.base64 = 0;
	}

	/* Args pass 2 */
	optind = 1;
	bs.base64 = (bs.style == STYLE_BASE64) ? 1 : 0;
	while ((fl = getopt(argc, argv, optstr)) != -1) {
		switch (fl) {
		case 'a':
			while ((cp = strsep(&optarg, ",")) != NULL) {
				if (*cp == '\0')
					continue;
				a_opts++;
				hf = bsdsum_op_find_alg(cp, bs.base64, 0);
				if (hf == NULL) 
					errx(1, "unsupported algorithm");
				if ((bs.style == STYLE_GNU ||
					bs.style == STYLE_CKSUM) &&
					( ! (hf->style & STYLE_SPACE) ||
					(hf->split >= 2)))
					errx(1, "style not supported for %s",
						cp);
				hf = bsdsum_add_op(&bs, hf, bs.enc64);
				if (hf->split >= 2)
					use_split = 1;
				hf->style = bs.style | hf->use_style;
			}
			break;
		case 'o':
			bs.ofile = open(optarg, O_WRONLY|O_CREAT|O_APPEND,
					0600);
			if (bs.ofile < 0)
				errx(1, "could not open %s", optarg);
			break;
		case 'C':
			bs.selective_checklist = optarg;
			break;
		case 'c':
			bs.cflag = 1;
			break;
		case 'p':
			bs.pflag = 1;
			break;
		case 't':
			bsdsum_autotest();
			break;
		case 'f':
		case 'l':
		case 's':
			/* already processed */
			break;
		default:
			bsdsum_help();
			exit(1);
		}
	}
	argc -= optind;
	argv += optind;

	if (bs.ofile < 0)
		bs.ofile = 1;

	/* Most arguments are mutually exclusive */
	if ((argc != 1) && (bs.length >= 0 || bs.offset >= 0))
		errx(1, "one file (and only one) must be specified with -f/-l");
	fl = bs.pflag + bs.cflag;
	if (fl > 1 || (fl && argc && bs.cflag == 0) || 
	    (bs.selective_checklist != NULL && argc == 0))
		errx(1, "non-compatible options on command-line");
	if (bs.selective_checklist || bs.cflag) {
		if (a_opts > 1)
			errx(1, "only one algorithm may be specified "
			    "in -C or -c mode");
	}

	/* missing alg name */
	if ((bsdsum_count_op(&bs) == 0) && ! bs.cflag && 
		! bs.selective_checklist) 
		errx(1, "missing algorithm name (-a)");

	/* split mode constrainsts */
	if (use_split) {
		if (bs.pflag || argc == 0)
			errx(1, "split digest not usable on stdin");
	}
	
	/* -c/-C */
	if (bs.selective_checklist) {
		int i;

		bs.error = bsdsum_digest_filelist(&bs, bs.selective_checklist, 
							bs.hl, argc, argv);
		for (i = 0; i < argc; i++) {
			if (argv[i] != NULL) {
				warnx("%s does not exist in %s", argv[i],
				    bs.selective_checklist);
				bs.error++;
			}
		}
	} 
	else if (bs.cflag) {
		if (argc == 0)
			bs.error = bsdsum_digest_filelist(&bs, "-", bs.hl,
								0, NULL);
		else {
			while (argc--) {
				bs.error += bsdsum_digest_filelist(&bs, *argv++,
			  					bs.hl, 0, NULL);
			}
		}
	} 
	else {
		if (bs.pflag || argc == 0)
			bs.error = bsdsum_digest_file(&bs, "-", bs.pflag);
		else {
			for (i = 0; i < argc; i++) 
				bs.error += bsdsum_digest_file(&bs, argv[i], 0);
		}
	}

	close(bs.ofile);
	return(bs.error ? EXIT_FAILURE : EXIT_SUCCESS);
}


